File: Problemset/1492-The kth Factor of n/Solution.rb
# @param {Integer} n
# @param {Integer} k
# @return {Integer}
def kth_factor(n, k)
    factor = 1
    i = 0

    while factor * factor <= n
        if n % factor == 0
            i += 1
            return factor if i == k
        end
        factor += 1
    end

    factor -= 1
    factor -= 1 if factor * factor == n

    while factor > 0
        if n % factor == 0
            i += 1
            return n / factor if i == k
        end
        factor -= 1
    end

    return -1
end

File: Problemset/1492-The kth Factor of n/Solution.py
class Solution:
    def kthFactor(self, n: int, k: int) -> int:
        factor = 1
        i = 0

        while factor * factor <= n:
            if n % factor == 0:
                i += 1
                if i == k:
                    return factor
            factor += 1

        factor -= 1
        factor -= 1 if factor * factor == n else 0

        while factor > 0:
            if n % factor == 0:
                i += 1
                if i == k:
                    return n // factor
            factor -= 1

        return -1
